File: C/transform.c
#include "transform.h"

transform_t transform_build(unsigned int *obj_buffer)
{
    transform_t transform;

    obj_buffer[3] = (unsigned int)malloc(3 * sizeof(float));
    obj_buffer[4] = (unsigned int)malloc(3 * sizeof(float));
    obj_buffer[5] = (unsigned int)malloc(3 * sizeof(float));

    transform.position = (float *)obj_buffer[3];
    transform.rotation = (float *)obj_buffer[4];
    transform.scale = (float *)obj_buffer[5];

    return transform;
}

File: web/src/Transform.js
import { vec3 } from "./math.js";

export default class Transform {
  #pos;
  #rot;
  #scl;

  constructor(pos = vec3(), rot = vec3(), scl = vec3(1, 1, 1)) {
    this.#pos = pos;
    this.#rot = rot;
    this.#scl = scl;
  }

  initialize() {
    this.position = this.#pos;
    this.rotation = this.#rot;
    this.scale = this.#scl;
  }

  get position() {
    return this.#pos;
  }
  set position(pos) {
    this.#pos = pos;
  }
  get rotation() {
    return this.#rot;
  }
  set rotation(rot) {
    this.#rot = rot;
  }
  get scale() {
    return this.#scl;
  }
  set scale(scl) {
    this.#scl = scl;
  }
}
